src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/BigQuerySQLEngine.java

```diff
@@ -0,0 +1,330 @@
+    BigQueryJoinDataset joinDataset = BigQueryJoinDataset.getInstance(sqlJoinRequest,
+                                                                      getStageNameToBQTableNameMap(),
+                                                                      bigQuery,
+                                                                      project,
+                                                                      dataset,
+                                                                      runId);
+
+    datasets.put(sqlJoinRequest.getDatasetName(), joinDataset);
+
+    return joinDataset;
+  }
+
+  @Override
+  public void cleanup(String datasetName) throws SQLEngineException {
+    BigQuerySQLDataset bqDataset = datasets.get(datasetName);
+    if (bqDataset == null) {
+      return;
+    }
+
+    cancelJob(datasetName, bqDataset);
+    deleteTable(datasetName, bqDataset);
+    deleteTempFolder(bqDataset);
+  }
+
+  /**
+   * Get a map that contains stage names as keys and BigQuery tables as Values.
+   *
+   * @return map representing all stages currently pushed to BQ.
+   */
+  protected Map<String, String> getStageNameToBQTableNameMap() {
+    return datasets.entrySet()
+      .stream()
+      .collect(Collectors.toMap(
+        Map.Entry::getKey,
+        e -> e.getValue().getBigQueryTableName()
+      ));
+  }
+
+  /**
+   * Stops the running job for the supplied dataset
+   *
+   * @param stageName the name of the stage in CDAP
+   * @param bqDataset the BigQuery Dataset Instance
+   */
+  protected void cancelJob(String stageName, BigQuerySQLDataset bqDataset) {
+    String jobId = bqDataset.getJobId();
+
+    // If this dataset does not specify a job ID, there's no need to cancel any job
+    if (jobId == null) {
+      return;
+    }
+
+    try {
+      String tableName = bqDataset.getBigQueryTableName();
+      Job job = bigQuery.getJob(jobId);
+
+      if (job == null) {
+        return;
+      }
+
+      if (!job.cancel()) {
+        LOG.error("Unable to cancel BigQuery job '{}' for table '{}}' and stage '{}'", jobId, tableName, stageName);
+      }
+    } catch (BigQueryException bqe) {
+      // Log exception and continue execution.
+      LOG.error("Exception when cancelling BigQuery job", bqe);
+    }
+  }
+
+  /**
+   * Deletes the BigQuery table for the supplied dataset
+   *
+   * @param stageName the name of the stage in CDAP
+   * @param bqDataset the BigQuery Dataset Instance
+   */
+  protected void deleteTable(String stageName, BigQuerySQLDataset bqDataset) {
+    try {
+      String tableName = bqDataset.getBigQueryTableName();
+      TableId tableId = TableId.of(project, dataset, tableName);
+
+      if (!bigQuery.delete(tableId)) {
+        LOG.error("Unable to delete BigQuery table '{}' for stage '{}'",  tableName, stageName);
+      }
+    } catch (BigQueryException bqe) {
+      // Log exception and continue execution.
+      LOG.error("Exception when deleting BigQuery table", bqe);
```

Comment:
doesn't seem right to swallow this exception. The cleanup() method declares that it can throw a SQLEngineException, but these methods log errors and move on. 

Would be better to propagate this exception up and have the cleanup() method catch and log/handle appropriately.

Reply:
I have re-factored the error handling when executing the cleanup step, PTAL.